image-rs/src/layer_store.rs

```diff
@@ -0,0 +1,79 @@
+use std::fs;
+use std::path::PathBuf;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// The LayerStore abstracts the image layers storage on the file system.
+pub struct LayerStore {
+    /// OCI image layer data store dir.
+    pub data_dir: PathBuf,
+
+    /// Next layer index
+    layers_index: AtomicUsize,
+}
+
+impl LayerStore {
+    /// Retrieves the next layer index by computing the maximum index from the
+    /// existing layers in data dir.
+    pub fn get_layer_index(data_dir: PathBuf) -> AtomicUsize {
+        let mut next = 0;
+        if data_dir.exists() {
+            if let Ok(paths) = fs::read_dir(data_dir) {
+                for entry_path in paths.filter_map(|result| result.ok()) {
+                    if let Some(name) = entry_path.file_name().to_str() {
+                        let n = name.parse::<usize>().unwrap_or(0);
```

Comment:
do we expect a parsing error here?

Reply:
Similar to above. If there is a parsing error, the file name is not an integer. It should not conflict with real layers.

Reply:
Added error handling as suggest by @mkulke